src/consensus/vote.cpp

```diff
@@ -309,4 +315,162 @@ bool VoteManager::pbftBlockHasEnoughValidCertVotes(PbftBlockCert const& pbft_blo
 }
 
+NextVotesForPreviousRound::NextVotesForPreviousRound(addr_t node_addr, std::shared_ptr<DbStorage> db)
+    : db_(db), enough_votes_for_null_block_hash_(false), voted_value_(blk_hash_t(0)), next_votes_size_(0) {
+  LOG_OBJECTS_CREATE("NEXT_VOTES");
+}
+
+void NextVotesForPreviousRound::clear() {
+  uniqueLock_ lock(access_);
+  enough_votes_for_null_block_hash_ = false;
+  voted_value_ = blk_hash_t(0);
+  next_votes_size_ = 0;
+  next_votes_.clear();
+}
+
+bool NextVotesForPreviousRound::haveEnoughVotesForNullBlockHash() const {
+  sharedLock_ lock(access_);
+  return enough_votes_for_null_block_hash_;
+}
+
+blk_hash_t NextVotesForPreviousRound::getVotedValue() const {
+  sharedLock_ lock(access_);
+  return voted_value_;
+}
+
+std::vector<Vote> NextVotesForPreviousRound::getNextVotes() {
+  std::vector<Vote> next_votes_bundle;
+
+  sharedLock_ lock(access_);
+  for (auto const& blk_hash_nv : next_votes_) {
+    for (auto const& v : blk_hash_nv.second) {
+      next_votes_bundle.emplace_back(v);
+    }
+  }
+  assert(next_votes_bundle.size() == next_votes_size_);
+
+  return next_votes_bundle;
+}
+
+size_t NextVotesForPreviousRound::getNextVotesSize() const {
+  sharedLock_ lock(access_);
+  return next_votes_size_;
+}
+
+void NextVotesForPreviousRound::setNextVotesSize(size_t const size) {
+  uniqueLock_ lock(access_);
+  next_votes_size_ = size;
+}
+
+// Assumption is that all votes are validated, in next voting phase, in the same round and step
+void NextVotesForPreviousRound::update(std::vector<Vote> const& next_votes, size_t const TWO_T_PLUS_ONE) {
+  LOG(log_nf_) << "There are " << next_votes.size() << " next votes for updating.";
+  if (next_votes.empty()) {
+    return;
+  }
+
+  clear();
+
+  {
+    upgradableLock_ lock(access_);
+    // Copy all next votes
+    for (auto const& v : next_votes) {
+      LOG(log_dg_) << "Add next vote: " << v;
+
+      auto voted_block_hash = v.getBlockHash();
+      if (next_votes_.count(voted_block_hash)) {
+        upgradeLock_ locked(lock);
+        next_votes_[voted_block_hash].emplace_back(v);
+      } else {
+        std::vector<Vote> votes{v};
+        upgradeLock_ locked(lock);
+        next_votes_[voted_block_hash] = votes;
+      }
+    }
+  }
+
+  auto next_votes_size = next_votes.size();
+
+  if (TWO_T_PLUS_ONE == 0) {
+    // Next votes from own DB. Don't need compare with 2t+1
+    assert(next_votes_.size() == 1 || next_votes_.size() == 2);
+    setNextVotesSize(next_votes_size);
+    return;
+  }
+
+  // Protect for malicious players. If no malicious players, will include either/both NULL BLOCK HASH and a non NULL
+  // BLOCK HASH
+  LOG(log_nf_) << "PBFT 2t+1 is " << TWO_T_PLUS_ONE;
+  {
+    upgradableLock_ lock(access_);
+    auto it = next_votes_.begin();
+    while (it != next_votes_.end()) {
+      if (it->second.size() >= TWO_T_PLUS_ONE) {
+        LOG(log_nf_) << "Voted PBFT block hash " << it->first << " has " << it->second.size() << " next votes";
```

Comment:
Should we not add round and step to this log message to be more useful?

Reply:
This is vote model, doesn't know the current round and step. The current round and step are in the pbft manager model.
And for the vote itself, has print the vote whole info(include voted round and step info) in the above.